### PR TITLE
Online edits

### PR DESCRIPTION
**Pull request by farhadrclass**
The first round, 
I am adding my edits and notes into next push

### CODE REVIEW

**Comment by paraynaud** [comment]
@farhadrclass I've allowed the generation of the .pdf (at least for your last commits).
You can take a look at the final result !

**Comment by farhadrclass** [comment]
@dpo @paraynaud  I added the examples of the FluxNLPModels with R2

File: paper/paper.md [hunk 1]
```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
...
+  # One-hot-encode the labels
+  ytrain, ytest = onehotbatch(ytrain, 0:9), onehotbatch(ytest, 0:9)
+
+  # Create DataLoaders (mini-batch iterators)
+  train_loader = DataLoader((xtrain, ytrain), batchsize = args.batchsize, shuffle = true)
+  test_loader = DataLoader((xtest, ytest), batchsize = args.batchsize)
+  @info "The type is " typeof(xtrain)
+  return train_loader, test_loader
+end
+```
```
Comment (paraynaud) [inline paper/paper.md:253]: This should be common with the KnetNLPModels.jl example

**Review by paraynaud — COMMENTED** [review]
Thanks for the work @farhadrclass.

Defining a neural network is a lot easier with Flux.
We should put the Knet.jl architecture and KnetNLPModel initialization into Annex, and make sure the code running for the remaining FluxNLPModel works also for the KnetNLPModel in Annex.

File: paper/paper.md [hunk 2]
```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
     )
 
 final_accuracy = KnetNLPModels.accuracy(LeNetNLPModel)
+```
+## FluxNLPModels.jl
+We will define the neural network architecture using Flux.jl, specifically the LeNet architecture [@lecun1998gradient]. First, we define some methods to assist with calculations of accuracy, data loading, and training arguments such as the number of epochs and batch size. These methods are inspired by Flux-zoo [@flux_model_zoo].
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:212]: Good catch, 
I am pushing my bib file

File: paper/paper.md [hunk 3]
```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
...
+  for (x, y) in data_loader
+    x, y = device(x), device(y)
+    ŷ = model(x)
+    ls += loss(ŷ, y, agg = sum)
+    acc += sum(onecold(ŷ) .== onecold(y)) ## Decode the output of the model
+    num += size(x)[end]
+  end
+  return ls / num, acc / num
+end
+```
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:281]: true, 
in Knet we have accuracy, so in Flux we need this to calculate the accuracy,
I will add one line to show

File: paper/paper.md [hunk 4]
```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
...
+   ```julia 
+      @kwdef mutable struct Args
+        batchsize::Int = 128    # batch size
+        epochs::Int = 10        # number of epochs
+        use_cuda::Bool = true   # use gpu (if cuda and gpu available)
+        verbose_freq = 10       # logging for every verbose_freq iterations
+      end
+      args = Args() # collect options in a struct for convenience
+      
+  ```
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:294]: I am not sure, 
It shows how Flux define it parameters

File: paper/paper.md (hunk 1 above)
Comment (farhadrclass) [reply paper/paper.md:255]: Sure, we can 
But here we create the data loader in the function as well 

I will change mine to be a bit different and @paraynaud  can you modify your code as well to use the function I make?

Comment (farhadrclass) [reply paper/paper.md:255]: I fixed it 
I will push this

**Comment by farhadrclass** [comment]
@paraynaud  ready for review

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:276]: but the accuracy and loss are problem related, here we assume the output are onehot encoding, that is why I think it would be much better to have it as user defined (similar to what Flux.jl did )
what do you think?

File: paper/paper.md [hunk 5]
```diff
@@ -108,14 +112,23 @@ Accordingly to the architecture, we chose the MNIST dataset [@lecun-bouttou-beng
...
+  # Reshape Data in order to flatten each image into a linear array
+  xtrain = Flux.flatten(xtrain)
+  xtest = Flux.flatten(xtest)
+
+  # One-hot-encode the labels
+  ytrain, ytest = onehotbatch(ytrain, 0:9), onehotbatch(ytest, 0:9)
+  return xtrain, ytrain, xtest, ytest
+end
 
+xtrn, ytrn, xtst, ytst = getdata(args)
```
Comment (paraynaud) [inline paper/paper.md:131]: Is this working with KnetNLPModels?

If not, make a function 
```julia
function get_MNIST()
  xtrain, ytrain = MLDatasets.MNIST(Tx = T, split = :train)[:]
  xtest, ytest = MLDatasets.MNIST(Tx = T, split = :test)[:]
end
```

**Comment by farhadrclass** [comment]
> @farhadrclass I've allowed the generation of the .pdf (at least for your last commits). You can take a look at the final result in details of the JOSS/.... job, then click on summary and look at the artifact !

I can not find this

File: paper/paper.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md:289]: I don't see where need Args require being a structure.
For the sake of the simplicity of the example, variables well-named are clearer.

File: paper/paper.md (hunk 1 above)
Comment (paraynaud) [reply paper/paper.md]: I will move and adapt the KnetNLPModel example when this PR is rebased.

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md:276]: What does `NLPModels.obj` returns?

For me, `loss_and_accuracy` complicates the example while it could be in FluxNLPModel.
You told be, you want to potentially change the loss function for a FluxNLPModel, does that mean you will have to define a `loss_and_accuracy` for each loss function you use?

If this is the case, I think it is wrong.

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:276]: So **NLPModels.obj** is accuracy on the mini-batch, in ML when we say accuracy it has different meaning (not sure they use it right but..) basically it is the accuracy on the accuracy set (here is the test data ) and their loss is also on the whole datasets nit just the current mini-batch

File: paper/paper.md (hunk 1 above)
Comment (farhadrclass) [reply paper/paper.md]: I edited the example

File: paper/paper.md [hunk 6]
```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
...
+  if next === nothing #one epoch is finished
+    @info "Epoch", data.epoch
+    data.i = 0
+    data.epoch += 1
+    return
+  end
+  (item, data.state) = next
+  nlp.current_training_minibatch = device(item) # move to cpu or gpu
+end
+```
```
[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: I don't understand why you don't have a method similar to `KnetNLPModels.minibatch_next_train!(nlp)` very similar to your callback that callback could call.

It only requires keeping `state` in the FluxNLPModel.

File: paper/paper.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:289]: I see your point, I can remove it but need to adjust all the method with the inputs, give me 10 min and I will update

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md:276]: `NLPModels.obj(::FluxNLPModel, x)` should be the loss function on the current training minibatch, this is what the optimization methods from JSOSolvers will use for the training.

File: paper/paper.md (hunk 5 above)
Comment (farhadrclass) [reply paper/paper.md:131]: my code should work, however the getData make sure we turn the y into onehot encoding as required by the model

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:276]: that is exactly what that is, NLPModels.obj(::FluxNLPModel, x) is  the loss function on the current training minibatch

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md:276]: You can compute accuracy on any dataset (wether training or test), but for new user `accuracy` using the test dataset by default (though an optional argument maybe) is fine.

File: paper/paper.md (hunk 5 above)
Comment (paraynaud) [reply paper/paper.md:131]: The KnetNLPModels consider `y` as an integer and not an onehot encoding

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md:276]: yes but I don't think we should handle it in the fluxNLPModel but the user can define their own method, This allow them to use the same method for Flux and FluxNLPModels and lower the methods in the FluxNLPModels

File: paper/paper.md (hunk 5 above)
Comment (farhadrclass) [reply paper/paper.md:131]: this is what used before
`ytrn[ytrn.==0] .= output_classes ` this is one hot encoding, all classes are zero, except the current class

Comment (paraynaud) [reply paper/paper.md]: No  `ytrn[ytrn.==0] .= output_classes` turns every integer from `ytrn` whose value is 0 into `outputclasses` who's valuing 10.

example:
```latex
output_classes = 10 
ytrn = [0,1,5,0]
ytrn[ytrn.==0] .= output_classes
ytrn == [10,1,5,10] #true
```

Comment (farhadrclass) [reply paper/paper.md]: my mistake, 
I changed the code based on your example

File: paper/paper.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: > This allow them to use the same method for Flux and FluxNLPModels and lower the methods in the FluxNLPModels

This is not a receivable argument, it downgrades the features of FluxNLPModels.
You want to work only on the FluxNLPModel and not on both `FluxNLPModel` and Flux architecture.

```julia
accuracy(nlp; data_loader=nlp.test_minibatch_iterator) = ...
```
solves your concerns, you can switch the `data_loader` to anything you want.

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: but if they already have define their accuracy method in Flux and want to integrade to FluxNLPModels?

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: Have you an example to share?

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: sure, 
Eventually in my research we want to work on transformers, here is their code https://github.com/chengchingwen/Transformers.jl/tree/master/src

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: https://github.com/chengchingwen/Transformers.jl/blob/bc393c9c7407a0a9ee5094c4b3a19ec12ab62aa6/example/BERT/mnli/train.jl#L36

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: or another one 
 accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: > Eventually in my research we want to work on transformers,

The transformer loss function is basic, and doesn't depend on the dataset nor the neural network.

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: > or another one accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])

Do you have a link?

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: > > or another one accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])
> 
> Do you have a link?

https://github.com/FluxML/model-zoo/blob/31e147faef61e7f26dc2d6124afbac0cfe518b6a/text/lang-detection/model.jl#L132

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: > 

The idea is that we have different accuracy functions, that user define, the simplicity of them is another discussion

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: Yes, but requiring the user to define a function to check the accuracy of a neural network is wrong, especially if a generic method working in most cases can be written.
At worst, the `accuracy` method fails, and in such case the user must write its own accuracy method.

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: I do not agree with that, especially that I do not find any other DNN library to code the accuracy as a method in them 
Flux.jl, pytorch from python. 
I am not seeing the benefit of having it in the model

[the remark this answers is not in the document]
Comment (paraynaud) [reply paper/paper.md]: Two benefits:
- the user don't have to implement this function, something like 
```julia
flux_accuracy(x, y) = mean(Flux.onecold(flux_model(x), classes) .== y);
```
should be generic enough to be applied in most cases;
- the example of the JOSS paper is simpler.

Without a `accuracy` method, you can directly define your FluxNLPModel, train it and not check the result of the training, I find it quite strange.

[the remark this answers is not in the document]
Comment (farhadrclass) [reply paper/paper.md]: added

**Comment by farhadrclass** [comment]
This is from: https://github.com/farhadrclass/KnetNLPModels.jl/pull/1

Straightforward pass onto Summary + Statement of need + KnetNLPModel example.

What is left to do (beside FluxNLPModel example), e.g. comment not answered by the changes :

mention : "It is important to stress that we want to be able to experiment with solvers that enforce a decrease of the merit function";
answer : "Why not just write solvers directly in Flux.jl or Knet.jl";
mention callback : but it appears in the example section.
After rebasing onto PR-JOSS-Flux and PR-JOSS:

move Knet model definition to the end to put Flux architecture on the front (with a mention to the Knet model);
make sure the code calling solver works for both models (same instruction, same callback ...)

**Comment by farhadrclass** [comment]
I added:

1. Example from FluxNLPModels
  1.1 Add accuracy to FluxNLPModels
  1.2 Added MinibatchNext to FluxNLPModels
  1.3 Simplify and clean the example
2. Write the example in Markdown and edit 
3. Move the text (Flux before Knet)
4. Edit the first section 
5. Edit the text overall
6. Apply @dpo  remarks 
7. Answer Why not just write solvers directly in Flux.jl or Knet.jl"; but I am not 100% sure if that is enough

@paraynaud  please review and let me know what you think?

**Comment by paraynaud** [comment]
Nice work, thanks!
